cookbook/examples/v2/streamlit_apps/agentic_rag_v2/app.py

```diff
@@ -0,0 +1,404 @@
+    # Track loaded URLs and files in session state
+    ####################################################################
+    if "loaded_urls" not in st.session_state:
+        st.session_state.loaded_urls = set()
+    if "loaded_files" not in st.session_state:
+        st.session_state.loaded_files = set()
+    if "knowledge_base_initialized" not in st.session_state:
+        st.session_state.knowledge_base_initialized = False
+
+    st.sidebar.markdown("#### 📚 Document Management")
+    input_url = st.sidebar.text_input("Add URL to Knowledge Base")
+    if (
+        input_url and not prompt and not st.session_state.knowledge_base_initialized
+    ):  # Only load if KB not initialized
+        if input_url not in st.session_state.loaded_urls:
+            alert = st.sidebar.info("Processing URLs...", icon="ℹ️")
+            if input_url.lower().endswith(".pdf"):
+                try:
+                    # Download PDF to temporary file
+                    response = requests.get(input_url, stream=True, verify=False)
+                    response.raise_for_status()
+
+                    with tempfile.NamedTemporaryFile(
+                        suffix=".pdf", delete=False
+                    ) as tmp_file:
+                        for chunk in response.iter_content(chunk_size=8192):
+                            tmp_file.write(chunk)
+                        tmp_path = tmp_file.name
+
+                    reader = PDFReader()
+                    docs: List[Document] = reader.read(tmp_path)
+
+                    # Clean up temporary file
+                    os.unlink(tmp_path)
+                except Exception as e:
+                    st.sidebar.error(f"Error processing PDF: {str(e)}")
+                    docs = []
+            else:
+                scraper = WebsiteReader(max_links=2, max_depth=1)
+                docs: List[Document] = scraper.read(input_url)
+
+            if docs:
+                agentic_rag_agent.knowledge.load_documents(docs, upsert=True)
+                st.session_state.loaded_urls.add(input_url)
+                st.sidebar.success("URL added to knowledge base")
+            else:
+                st.sidebar.error("Could not process the provided URL")
+            alert.empty()
+        else:
+            st.sidebar.info("URL already loaded in knowledge base")
+
+    uploaded_file = st.sidebar.file_uploader(
+        "Add a Document (.pdf, .csv, or .txt)", key="file_upload"
+    )
+    if (
+        uploaded_file and not prompt and not st.session_state.knowledge_base_initialized
+    ):  # Only load if KB not initialized
+        file_identifier = f"{uploaded_file.name}_{uploaded_file.size}"
+        if file_identifier not in st.session_state.loaded_files:
+            alert = st.sidebar.info("Processing document...", icon="ℹ️")
+            file_type = uploaded_file.name.split(".")[-1].lower()
+            reader = get_reader(file_type)
```

Comment:
You shouldn't need the reader? That is like the whole benefit of the new knowledge class